src/scriptdev2/scripts/outland/shadowmoon_valley.cpp

```diff
@@ -1217,65 +1218,55 @@ enum
+        switch (pSummoned->GetEntry())
         {
-            // Only when it's close to the totem
-            if (!pWho->IsWithinDistInMap(m_creature, 1.5f))
-                return;
-
-            switch (uiEntry)
-            {
-                case NPC_EARTHEN_SOUL:
-                    pWho->CastSpell(m_creature, SPELL_EARTH_CAPTURED, TRIGGERED_OLD_TRIGGERED);
-                    break;
-                case NPC_FIERY_SOUL:
-                    pWho->CastSpell(m_creature, SPELL_FIERY_CAPTURED, TRIGGERED_OLD_TRIGGERED);
-                    break;
-                case NPC_WATERY_SOUL:
-                    pWho->CastSpell(m_creature, SPELL_WATER_CAPTURED, TRIGGERED_OLD_TRIGGERED);
-                    break;
-                case NPC_AIRY_SOUL:
-                    pWho->CastSpell(m_creature, SPELL_AIR_CAPTURED, TRIGGERED_OLD_TRIGGERED);
-                    break;
-            }
-
-            // Despawn the spirit soul after it's captured
-            ((Creature*)pWho)->ForcedDespawn();
+            case NPC_EARTHEN_SOUL:
+                pSummoned->CastSpell(m_creature, SPELL_EARTH_CAPTURED, TRIGGERED_OLD_TRIGGERED);
+                break;
+            case NPC_FIERY_SOUL:
+                pSummoned->CastSpell(m_creature, SPELL_FIERY_CAPTURED, TRIGGERED_OLD_TRIGGERED);
+                break;
+            case NPC_WATERY_SOUL:
+                pSummoned->CastSpell(m_creature, SPELL_WATER_CAPTURED, TRIGGERED_OLD_TRIGGERED);
+                break;
+            case NPC_AIRY_SOUL:
+                pSummoned->CastSpell(m_creature, SPELL_AIR_CAPTURED, TRIGGERED_OLD_TRIGGERED);
+                break;
         }
-    }
-
-    void OwnerKilledUnit(Unit* pVictim) override
-    {
-        if (pVictim->GetTypeId() != TYPEID_UNIT)
-            return;
-
-        uint32 uiEntry = pVictim->GetEntry();
 
-        // make elementals cast the sieve is only way to make it work properly, due to the spell target modes 22/7
-        if (uiEntry == NPC_EARTH_SPIRIT || uiEntry == NPC_FIERY_SPIRIT || uiEntry == NPC_WATER_SPIRIT || uiEntry == NPC_AIR_SPIRIT)
-            pVictim->CastSpell(pVictim, SPELL_ELEMENTAL_SIEVE, TRIGGERED_OLD_TRIGGERED);
+        // Despawn the spirit soul after it's captured
+        ((Creature*)pSummoned)->ForcedDespawn(1000);
     }
 
     void JustSummoned(Creature* pSummoned) override
     {
         // After summoning the spirit soul, make it move towards the totem
-        float fX, fY, fZ;
-        m_creature->GetContactPoint(pSummoned, fX, fY, fZ);
-        pSummoned->GetMotionMaster()->MovePoint(1, fX, fY, fZ);
+        pSummoned->GetMotionMaster()->MovePoint(1, m_creature->GetPositionX(), m_creature->GetPositionY(), m_creature->GetPositionZ() + 4);
+    }
+
+    void UpdateAI(const uint32 uiDiff) override
+    {
+        if (m_uiElementalSieveTimer <= uiDiff)
+        {
+            m_creature->CastSpell(m_creature, SPELL_ELEMENTAL_SIEVE, TRIGGERED_OLD_TRIGGERED);
```

Comment:
We will need to make a thread about that I think. You are right for wanting to use it, because all the other code is used for this, but I dont want to because its not a combat spell, i.e. all code inside DoCastSpellIfCan is reduntant for this particular script (look at the checks inside). Also, I found that this function fails in the most unfortunate situations, causing a script here and there to malfunction. I guess after we write Cooldown API, we will need to deal with it anyway.